### ISSUE TITLE
FR: ability to reorder the channel order when using 4 decks 

### ISSUE THREAD

**Issue by dmt-hub** [mixxxdj/mixxx#13977]
### Feature Description

at the moment when using 4 decks it places deck 1 & 2 in the middle of the mixer so it goes like this: 3 1 2 4 - which can be confusing.
would be great if we could have the order: 1 2 3 4 in the mixer section when using 4 decks.
<img width="228" alt="Screenshot 2024-12-04 at 20 22 59" src="https://github.com/user-attachments/assets/6ac931de-d238-4629-9734-c7c183927e0e">

**Comment by Swiftb0y** [comment]
Thats also a skin feature. And not a super hard one to implement at that. Take a look at our Skin documentation: https://github.com/mixxxdj/mixxx/wiki#skin-documentation I'm sure you'd be able to hack this feature together yourself.

**Comment by ronso0** [comment]
Yes, most of it is just how the decks are arranged in skins, but note that 3-1-2-4 is also assumed in some places of the engine, for example for the default crossfader assignments (1, 3 left | 2,4 right).
Though this can also be set in the skin's `<attributes>` section IINM.

**Comment by daschuer** [comment]
We have already a WIP PR for it: 
https://github.com/mixxxdj/mixxx/pull/13724

This is a duplicate of https://github.com/mixxxdj/mixxx/issues/11174
https://mixxx.zulipchat.com/#narrow/channel/247619-UI-.26-UX-design/topic/4.20Deck.20Waveform.20ordering

**Comment by dmt-hub** [comment]
@ronso0 
In which .xml file can i change these settings?

**Comment by dmt-hub** [comment]
In mixer_4decks.xml i changed it to:

          <Template src="skins:LateNight/mixer/channel_4decks.xml">
            <SetVariable name="ChanNum">1</SetVariable>
            <SetVariable name="xfaderLeft">default</SetVariable>
            <SetVariable name="xfaderRight">warning</SetVariable>
          </Template>

          <Template src="skins:LateNight/mixer/channel_4decks.xml">
            <SetVariable name="ChanNum">2</SetVariable>
            <SetVariable name="xfaderLeft">default</SetVariable>
            <SetVariable name="xfaderRight">warning</SetVariable>
          </Template>

          <Template src="skins:LateNight/mixer/channel_4decks.xml">
            <SetVariable name="ChanNum">3</SetVariable>
            <SetVariable name="xfaderLeft">warning</SetVariable>
            <SetVariable name="xfaderRight">default</SetVariable>
          </Template>

          <Template src="skins:LateNight/mixer/channel_4decks.xml">
            <SetVariable name="ChanNum">4</SetVariable>
            <SetVariable name="xfaderLeft">warning</SetVariable>
            <SetVariable name="xfaderRight">default</SetVariable>
          </Template>

But that doesnt seem to work, the channels stay the same

**Comment by dmt-hub** [comment]
@ronso0 
So i made a duplicate of LateNight skin residing here: /Applications/Mixxx.app/Contents/Resources/skins/LateNight copy
changed mixer_4decks.xml & waveforms_container.xml  
saved those files
reloaded the skin in mixxx
the channels are still in the same location 3 1 2 4

**Comment by dmt-hub** [comment]
@ronso0 
problem solved by renaming the folder to LateNight
Thanks for the help!

**Comment by ronso0** [comment]
> We have already a WIP PR for it: #13724
> 
> This is a duplicate of #11174 https://mixxx.zulipchat.com/#narrow/channel/247619-UI-.26-UX-design/topic/4.20Deck.20Waveform.20ordering

Thanks, that's the discussion I was looking for.
Not really a duplicate because there is still no possibility to reorder the channels, so let's keep this open.

**Comment by Fre0Grella** [comment]
Hello, I would like to work on this issue, but i don't really know if this could be a button or a switch on the option menu? Any suggestion for the best place to place this feature? @ronso0

**Comment by ronso0** [comment]
Thanks for the offer to work on this feature.
Though, IMHO this is not a starter / 'easy' feature as it requires quite some knowledge / reverse-engineering of the engine.
Also, besides the engine aspects of this (which I'm sure I'm no aware of myself), this is a really complex feature with the current QWidget-based skins and we should aim to get it in the QML skin.

I think this should be an option in the Mixer preferences btw.

**Comment by Fre0Grella** [comment]
Oook, that's not very reassuring but I'll try my best

**Comment by daschuer** [comment]
I think there is a chance to have this in the widget based skin as well. We have already the 2 to 4 deck changes "[Skin],show_4decks" . This will add up on that and can be reused in QML.

The engine does not mind where a certain deck is shown, but there is an issue with the Crossfader and Auto-DJ. 
Auto-DJ will automatically pick the first left and right deck. Normally 1 and 2.  After chaning deck orientation of decks 2 and 3 it will pick 1 and 3. Is that OK?  
https://github.com/mixxxdj/mixxx/blob/719833708c6e04004960f6be5faece32de6bb6bd/src/library/autodj/autodjprocessor.cpp#L382

Here a small draft to do list directly from my head (please adjust)
* Add a POC PR that changes to 1234 mode statically and test.  
* Add a preference option to the Interface preferences (Skins)  
* Implement a way to switch to the 1234 view. I think using a CO, something like `[Skin],deck_order_1234` will work.  
* Adjust default x-fader deck orientation "[ChannelX],orientation"

@ronso0 does that sound reasonable?

**Comment by ronso0** [comment]
Yes, that makes sense.
I was just unsure whether xfader orientation is the only aspect that needs to be considered.
But we'll know after
> * Add a POC PR that changes to 1234 mode statically and test.

Admittedly, the skin part is doable with waveform Singletons.

**Comment by Fre0Grella** [comment]
Sorry, for being late I hadn't check the comment on this issue for a while 😅
